### PR TITLE
Add multi-viewports for DX12

### PR DESCRIPTION
**Pull request by obfuscate**
So, at this moment viewports don't work with DX12 and it is my approach to solve the problem.
After this fix I had got the next result:
imgui_dx12_viewports

I hope it will be useful for someone apart from me :)

### CODE REVIEW

**Comment by obfuscate** [comment]
Urgh, ok, it doesn't still work for x86 applications.

**Comment by ocornut** [comment]
Thank you so much for that.. will look into it shortly.. 
Been long awaited. It seems surprisingly simpler than what it took to get the equivalent thing working for Vulkan.

(We can fix the 32-bit issue separately, it's been dangling for a while..)

File: examples/imgui_impl_dx12.cpp [hunk 1]
```diff
@@ -665,123 +729,221 @@ void ImGui_ImplDX12_NewFrame()
...
+            data->SwapChain->GetBuffer(i, IID_PPV_ARGS(&back_buffer));
+            g_pd3dDevice->CreateRenderTargetView(back_buffer, NULL, data->FrameCtx[i].RenderTargetCpuDescriptors);
+            data->FrameCtx[i].RenderTarget = back_buffer;
+        }
+    }
+
+    for (UINT i = 0; i < g_numFramesInFlight; i++)
+    {
+        if (data->Resources[i].IndexBuffer) { data->Resources[i].VertexBuffer->Release(); data->Resources[i].IndexBuffer = NULL; }
+        if (data->Resources[i].VertexBuffer) { data->Resources[i].VertexBuffer->Release(); data->Resources[i].VertexBuffer = NULL; }
```
Comment (ocornut) [inline examples/imgui_impl_dx12.cpp]: Oops error here!
I'll fix it in my copy (I'm making an additional patch to use SafeRelease everywhere in the patch)

**Comment by ocornut** [comment]
Hello @obfuscate,

Spotted an error (above).
However even with that fix, when I enabled DX12_ENABLE_DEBUG_LAYER 1 in main.cpp there are leaks being spotted. If you can find what the cause is let me know (I can report the fix in my second patch).

Thanks!

File: examples/imgui_impl_dx12.cpp [hunk 2]
```diff
@@ -46,16 +46,89 @@ static ID3D12Resource*              g_pFontTextureResource = NULL;
...
-static FrameResources*  g_pFrameResources = NULL;
+
 static UINT             g_numFramesInFlight = 0;
-static UINT             g_frameIndex = UINT_MAX;
+
+struct FrameContext
+{
+    ID3D12CommandAllocator*     CommandAllocator;
+    UINT64                      FenceValue;
+    ID3D12Resource*             RenderTarget;
```
Comment (ocornut) [inline examples/imgui_impl_dx12.cpp]: FenceValue is never used, would that be a problem?

**Comment by ocornut** [comment]
I found the leak, a missing `->Release()` on the IDXGISwapChain1 pointer in `ImGui_ImplDX12_CreateWindow()`, fixing now.

```
    // Or swapChain.As(&mSwapChain)
    swap_chain->QueryInterface(IID_PPV_ARGS(&data->SwapChain));
++  swap_chain->Release();
```

**Comment by ocornut** [comment]
Merged now, thank you @obfuscate for the excellent PR !
I will also push a change on master now to add the new parameter to Init() function so we can break this API earlier on.

One of the thing I have noticed is that dropping a viewported window back into the main viewport flickers (it seems like the viewport gets detroyed before the main viewport gets a chance to present a new picture). Any suggestion?

File: examples/imgui_impl_dx12.cpp (hunk 1 above)
Comment (obfuscate) [reply examples/imgui_impl_dx12.cpp]: Oh, sorry, my bad.

File: examples/imgui_impl_dx12.cpp (hunk 2 above)
Comment (obfuscate) [reply examples/imgui_impl_dx12.cpp]: Oh, It seems it was replaced by FenseSignaledValue and not FenceValue is redundant.

**Comment by obfuscate** [comment]
> One of the thing I have noticed is that dropping a viewported window back into the main viewport flickers (it seems like the viewport gets detroyed before the main viewport gets a chance to present a new picture). Any suggestion?
Yeah, Unfortunately we see flickering when a window return back to the main viewport. 
Can we merely add ImDrawLists from the destroyed window to the parent/main viewport to avoid flickering? In this case we lose the buffers of the destroyed window, but the parent/main window considers these data. Or isn't it impossible because of the order of calls (for example, RenderDrawData is called before DestroyWindow)?

**Comment by obfuscate** [comment]
@ocornut , I also have found one mistake in the code:

`IM_ASSERT(data->FrameCtx[0].RenderTarget == NULL && data->FrameCtx[1].RenderTarget == NULL && data->FrameCtx[2].RenderTarget == NULL);
`

It might lead to UB if there are for example only 2 circular frames. I can propose the next solutiond:
`for (int i = 0; i < g_numFramesInFlight; ++i)
	{
		IM_ASSERT(data->FrameCtx[i].RenderTarget == NULL);
	}`

**Comment by ocornut** [comment]
Hello,

> It might lead to UB if there are for example only 2 circular frames. I can propose the next solution:

I have fixed this.

Could you make a PR for the other fix (maybe calling the function `ImGui_ImplDX12_WaitForGPU`), I am not sure exactly where to do put and I would like the remove the `if (viewport != ImGui::GetMainViewport())` test (replace it with e.g. CommandQueue if the main viewport doesn't have a command queue?). Because we are slowly going to transition the concept of "main viewport" toward "0+ viewports owned by app" (often it will be 1) and "viewports managed by the back-end" so there won't be a single main viewport concept any more.

Thank you!

**Comment by obfuscate** [comment]
Sorry for the delay. I will try to create a PR this weekends.

**Pull request closed without merging**